### ISSUE TITLE
Incompatibility with Python 3.8 and 3.9: TypeError in file_store.py

### ISSUE THREAD

**Issue by similato87** [AntonOsika/gpt-engineer#898]
## Policy and info
- Maintainers will close issues that have been stale for 14 days if they contain relevant answers.
- Adding the label "sweep" will automatically turn the issue into a coded pull request. Works best for mechanical tasks. More info/syntax at: https://docs.sweep.dev/

## Expected Behavior
The project documentation states support for Python versions 3.8 - 3.11. I expect the software to run without syntax errors in these versions.

## Current Behavior
When attempting to run the project in Python 3.9, a TypeError occurs in `file_store.py` due to the use of the union operator `|` in type hints.

## Failure Information
The project uses a syntax feature (`str | Path`) that is only available in Python 3.10 and later, leading to incompatibility with Python 3.8 and 3.9.

### Steps to Reproduce
1. Set up the project in a Python 3.9 environment.
2. Follow the installation and setup instructions.
3. Attempt to run the project, leading to the TypeError in `file_store.py`.

### Failure Logs
```
Traceback (most recent call last):
  File ".../Scripts/gpt-engineer", line 3, in <module>
    from gpt_engineer.applications.cli.main import app
  ... (additional traceback)
  File ".../file_store.py", line 8, in FileStore
    def __init__(self, path: str | Path | None = None):
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

**Comment by ATheorell** [comment]
Thanks for reporting, we'll take a decision on either making the python requirement stricter or changing the syntax

**Comment by ErikBjare** [comment]
I think the right call is to just bump the minimum Python version to 3.10

**Comment by similato87** [comment]
### In the file [file_store.py](https://github.com/AntonOsika/gpt-engineer/blob/main/gpt_engineer/core/default/file_store.py) on line 9, there's a line of code: `def __init__(self, path: str | Path | None = None):`. This syntax, specifically the use of the `|` operator for type hints, causes a syntax error in Python versions 3.9 and below.

### A potential solution is to modify this line to use the `Union` type hint from the `typing` module: `def __init__(self, path: Union[str, Path, None] = None):`. This change should resolve the syntax error across all supported Python versions, including 3.9 and lower.


### Does this sound like an option?

**Comment by ATheorell** [comment]
@similato87 we are leaning towards bumping the python requirement. Do you have a strong argument why it is important that we support python 3.8 and 3.9?

**Comment by similato87** [comment]
Hi @ATheorell ! Here are some reasons for maintaining compatibility with Python 3.8 and 3.9 in the GPT Engineer project:

1. **Organizational Standard and Integration:** Python 3.8 and 3.9 are key in many organizations, especially where legacy systems and long-term stability are priorities. Supporting these versions ensures seamless integration with existing infrastructures.

2. **Increased Adoption:** Accommodating Python 3.8 and 3.9 can expand GPT Engineer's user base to include environments slow in adopting newer Python versions.

3. **Minimal Syntax Adjustments:** The only major compatibility issue is a syntax error in one part of the code for Python 3.9 and lower, which seems manageable to fix.

Additionally, I have expertise in Python 3.9 and am eager to contribute towards maintaining compatibility for these versions. My commitment extends to further supporting the project's overall development.

**Comment by similato87** [comment]
> Hi @ATheorell ! Here are some reasons for maintaining compatibility with Python 3.8 and 3.9 in the GPT Engineer project:
> 
> 1. **Organizational Standard and Integration:** Python 3.8 and 3.9 are key in many organizations, especially where legacy systems and long-term stability are priorities. Supporting these versions ensures seamless integration with existing infrastructures.
> 2. **Increased Adoption:** Accommodating Python 3.8 and 3.9 can expand GPT Engineer's user base to include environments slow in adopting newer Python versions.
> 3. **Minimal Syntax Adjustments:** The only major compatibility issue is a syntax error in one part of the code for Python 3.9 and lower, which seems manageable to fix.
> 
> Additionally, I have expertise in Python 3.9 and am eager to contribute towards maintaining compatibility for these versions. My commitment extends to further supporting the project's overall development.
## Issue with Type Hints Compatibility in Python 3.9

### Background
After facing compatibility challenges with Python 3.9, I changed about ten type hints in the `application` and `core` folders from the Python 3.10 style to the traditional form. This allowed the project to run successfully:
- ![1](https://github.com/AntonOsika/gpt-engineer/assets/71301573/fd513e12-c0da-4bd0-b95d-1f024cddca8a)


### Observation
I noticed the project uses mixed type hint styles, particularly in [`disk_memory.py`](https://github.com/AntonOsika/gpt-engineer/blob/main/gpt_engineer/core/default/disk_memory.py). Standardizing on a consistent type hint style, especially if Python 3.10 is the goal, is advisable for code consistency.

### Suggestion
Considering many in the machine learning field, including myself and colleagues, frequently use Python 3.9, extending support to earlier versions is recommended. The process of configuring `pipx` and `poetry` was cumbersome, especially when I had to switch to Python 3.11 for compatibility. Broader support for Python versions would greatly benefit users like us who prefer Python 3.9.

**Comment by ErikBjare** [comment]
Python 3.8 is EOL in 8 months and 3.9 in 21 months.

Both have ended their Active Support cycle: https://endoflife.date/python

Python 3.10 has been out for 2 years.

I vastly prefer working with a 3.10 codebase due to the syntax improvements. While there are certainly people/orgs slow to upgrade their toolchain, I'd hate for those slow adopters to drag everyone down.

While it's true that there's minimal work to support 3.8/3.9, they are both very much on the way out. We'd probably only support them for ~12 months anyway, and given the recent refactor I think now is a good time to rip the bandaid off.

**Comment by ATheorell** [comment]
Suggestion @similato87: If it's little work for you to adapt the syntax to be backwards compatible, can you make a PR with these changes?  

Being backwards compatible can't harm, even if we make version requirement stricter.

**Comment by ATheorell** [comment]
@ErikBjare can you improve the config of the linters?

**Comment by ErikBjare** [comment]
@ATheorell Done in #908. 

>Being backwards compatible can't harm

That's not entirely true. Dependencies will at some point drop support for Python 3.8 (if some haven't already), leaving us pinned with old versions.

>even if we make version requirement stricter

If version requirement is made stricter, there's no reason to use the old syntax.

`pyupgrade` should be run when the decision is eventually made to drop support for older versions. This would automatically update old syntax to the new one.

**Comment by similato87** [comment]
> Suggestion @similato87: If it's little work for you to adapt the syntax to be backwards compatible, can you make a PR with these changes?
> 
> Being backwards compatible can't harm, even if we make version requirement stricter.

Hi @ATheorell,

Following the [contributing guidelines](https://github.com/AntonOsika/gpt-engineer/blob/main/.github/CONTRIBUTING.md), I've completed the backward compatibility refactor on my local git. However, I encountered an issue with access denial when attempting to push the changes to the remote repository. Could you please grant the necessary permissions or suggest an alternative way to submit the changes?

**Comment by ErikBjare** [comment]
@similato87 You need to create a fork, push your changes to that fork, and submit a PR from that fork.

**Comment by similato87** [comment]
> @ATheorell Done in #908.
> 
> > Being backwards compatible can't harm
> 
> That's not entirely true. Dependencies will at some point (probably soon, given upcoming EOL) drop support for Python 3.8 (if some haven't already), leaving us pinned with old versions.
> 
> > even if we make version requirement stricter
> 
> If version requirement is made stricter, there's no reason to use the old syntax.
> 
> `pyupgrade` should be run when the decision is eventually made to drop support for older versions. This would automatically update old syntax to the new one.

@ErikBjare l, I concur with your insights regarding the trade-offs of backward compatibility. While supporting older Python versions is beneficial for wider accessibility, it does bring the challenge of dependency management. As you mentioned, dependencies will eventually drop support for older versions like Python 3.8, and this could limit us to outdated packages.

Adopting `pyupgrade` at the appropriate time is a forward-thinking strategy. In the interim, maintaining support for older versions could be a pragmatic approach, with the understanding that we'll transition as dependencies evolve. Regular releases could be planned to align with these dependency updates, ensuring a smooth transition for users.

**Comment by similato87** [comment]
I've completed the necessary changes and have created a PR(https://github.com/AntonOsika/gpt-engineer/pull/909). However, I noticed that the PR failed the pre-commit checks. While this concerns me, it's important to note that all my changes have been thoroughly tested and run successfully on my local machine. I'm wondering if this discrepancy could be due to environment-specific settings or configurations. I would appreciate any guidance or review to ensure the changes meet the project standards.
